compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,409 @@
+     *  ```scala
+     *  ${  (using Quotes) ?=> {... x ... X ...} }: T2
+     *  ```
+     *  is transformed into
+     * ```scala
+     *  {{{ <holeIdx++> | T2 | x, X | (x$1: Expr[T1], X$1: Type[X]) => (using Quotes) ?=> {... ${x$1} ...  X$1.Underlying ...}  }}}
+     *  ```
+     */
+    private class SpliceTransformer(spliceOwner: Symbol) extends Transformer:
+      private var refBindingMap = mutable.Map.empty[Symbol, (Tree, Symbol)]
+      /** Reference to the `Quotes` instance of the current level 1 splice */
+      private var quotes: Tree = null
+      private var healedTypes: PCPCheckAndHeal.QuoteTypeTags = null
+
+      def transformSplice(tree: tpd.Tree, tpe: Type)(using Context): tpd.Tree =
+        assert(level == 0)
+        val newTree = transform(tree)
+        val (refs, bindings) = refBindingMap.values.toList.unzip
+        val bindingsTypes = bindings.map(_.termRef.widenTermRefExpr)
+        val methType = MethodType(bindingsTypes, newTree.tpe)
+        val meth = newSymbol(spliceOwner, nme.ANON_FUN, Synthetic | Method, methType)
+        val ddef = DefDef(meth, List(bindings), newTree.tpe, newTree.changeOwner(ctx.owner, meth))
+        val fnType = defn.FunctionType(bindings.size, isContextual = false).appliedTo(bindingsTypes :+ newTree.tpe)
+        val closure = Block(ddef :: Nil, Closure(Nil, ref(meth), TypeTree(fnType)))
+        holeIdx += 1
+        tpd.Hole(true, holeIdx, refs, closure, TypeTree(tpe))
+
+      override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
+        tree match
+          case tree: RefTree =>
+            if tree.isTerm then
+              if quotedDefs.contains(tree.symbol) then
+                val tpe = tree.tpe.widenTermRefExpr match {
+                  case tpw: MethodicType => tpw.toFunctionType(isJava = false)
+                  case tpw => tpw
+                }
+                spliced(tpe)(capturedTerm(tree))
+              else super.transform(tree)
+            else // tree.isType then
+              if containsCapturedType(tree.tpe) then
+                val capturedTypeSym = capturedType(tree)
+                if level >= 1 then
+                  TypeTree(healedTypes.getTagRef(capturedTypeSym.termRef))
+                else
+                  ref(capturedTypeSym).select(defn.QuotedType_splice)
+              else super.transform(tree)
+          case tree: TypeTree =>
+            if containsCapturedType(tree.tpe) then
+              // TODO factor out the common code with the above case
+              val capturedTypeSym = capturedType(tree)
+              if level >= 1 then
+                TypeTree(healedTypes.getTagRef(capturedTypeSym.termRef))
+              else
+                ref(capturedTypeSym).select(defn.QuotedType_splice)
+            else tree
+          case tree @ Assign(lhs: RefTree, rhs) =>
+            if quotedDefs.contains(lhs.symbol) then transformSplicedAssign(tree)
+            else super.transform(tree)
```

Comment:
It can be scope extruded like any other expression 
```scala
var bad: Expr[Any] = null
'{
  var x = 2
  ${ bad = '{ x = 3 }; 'x }
}
bad
```
The usual runtime mechanism will catch these extrusions.

The assignment itself does not insert any other kind of extrusion.